wandb/sdk/internal/job_builder.py

```diff
@@ -0,0 +1,252 @@
+        name = make_artifact_name_safe(f"job-{image_name}")
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        source: ImageSourceDict = {
+            "image": image_name,
+            "args": args,
+        }
+        return artifact, source
+
+    def build(self) -> Optional[Artifact]:
+        if not os.path.exists(
+            os.path.join(self._settings.files_dir, REQUIREMENTS_FNAME)
+        ):
+            return None
+        metadata = self._handle_metadata_file()
+        if metadata is None:
+            return None
+        git_info: Dict[str, str] = metadata.get("git", {})
+        image_name: Optional[str] = metadata.get("docker", None)
+        program_relpath: Optional[str] = metadata.get("codePath", None)
+        args: List[str] = metadata.get("args", [])
+        runtime: Optional[str] = metadata.get("python", None)
+        # can't build a job without a python version
+        if runtime is None:
+            return None
+        artifact = None
+        source_type = None
+        source: Optional[
+            Union[GitSourceDict, ArtifactSourceDict, ImageSourceDict]
+        ] = None
+        if self._has_git_job_ingredients(git_info, program_relpath):
+            remote = git_info.get("remote")
+            commit = git_info.get("commit")
+            assert isinstance(remote, str)
+            assert isinstance(commit, str)
+            assert program_relpath is not None
+            artifact, source = self._build_repo_job(
+                remote, commit, program_relpath, args
+            )
+            source_type = "repo"
+        elif self._has_artifact_job_ingredients(program_relpath):
+            assert program_relpath is not None
+            artifact, source = self._build_artifact_job(program_relpath, args)
+            source_type = "artifact"
+        elif self._has_image_job_ingredients(image_name):
+            assert image_name is not None
+            artifact, source = self._build_image_job(image_name, args)
+            source_type = "image"
+
+        if artifact is None or source_type is None or source is None:
+            return None
+
+        input_types = TypeRegistry.type_of(self._config).to_json()
+        output_types = TypeRegistry.type_of(self._summary).to_json()
+
+        source_info: JobSourceDict = {
+            "_version": "v0",
+            "source_type": source_type,
+            "source": source,
+            "input_types": input_types,
+            "output_types": output_types,
+            "runtime": runtime,
+        }
+
+        with artifact.new_file("wandb-job.json") as f:
+            f.write(json.dumps(source_info, indent=4))
+
+        artifact.add_file(
+            os.path.join(self._settings.files_dir, REQUIREMENTS_FNAME),
+            name=FROZEN_REQUIREMENTS_FNAME,
+        )
+
+        return artifact
+
+    def _handle_metadata_file(
+        self,
+    ) -> Optional[Dict]:
+        # TODO: settings static is not populated with several
```

Comment:
is it just an issue that settings static is missing some of the fields?

Reply:
I should actually fix this comment, I believe we need to use the metadata file to get accurate values when a run is synced after being run in offline mode